File: ROS2/day1/hw2/tongsin_py/tongsin_py/my_subscriber.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String, Int32, Float32, Bool


class MySubscriber(Node):

    def __init__(self):
        super().__init__('my_py_subscriber')
        self.str_sub = self.create_subscription(
            String, 'my_str_topic', self.str_callback, 10)
        self.int_sub = self.create_subscription(
            Int32, 'my_int_topic', self.int_callback, 10)
        self.float_sub = self.create_subscription(
            Float32, 'my_float_topic', self.float_callback, 10)
        self.bool_sub = self.create_subscription(
            Bool, 'my_bool_topic', self.bool_callback, 10)

    def str_callback(self, msg):
        self.get_logger().info('Received string: {0}'.format(msg.data))

    def int_callback(self, msg):
        self.get_logger().info('Received int: {0}'.format(msg.data))

    def float_callback(self, msg):
        self.get_logger().info('Received float: {0}'.format(msg.data))

    def bool_callback(self, msg):
        self.get_logger().info('Received bool: {0}'.format(msg.data))


def main(args=None):
    rclpy.init(args=args)
    node = MySubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('Keyboard Interrupt (SIGINT)')
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ROS2/day1/hw2/tongsin/src/my_subscriber.cpp
#include "my_subscriber.hpp"

MySubscriber::MySubscriber() : Node("my_subscriber")
{
	my_str_subscriber_ = this->create_subscription<std_msgs::msg::String>(
		"my_str_topic",
		10,
		std::bind(&MySubscriber::topic_callback, this, std::placeholders::_1));
	my_int_subscriber_ = this->create_subscription<std_msgs::msg::Int32>(
		"my_int_topic",
		10,
		std::bind(&MySubscriber::int_callback, this, std::placeholders::_1));
	my_float_subscriber_ = this->create_subscription<std_msgs::msg::Float32>(
		"my_float_topic",
		10,
		std::bind(&MySubscriber::float_callback, this, std::placeholders::_1));
	my_bool_subscriber_ = this->create_subscription<std_msgs::msg::Bool>(
		"my_bool_topic",
		10,
		std::bind(&MySubscriber::bool_callback, this, std::placeholders::_1));
}

MySubscriber::~MySubscriber()
{
}

void MySubscriber::topic_callback(const std_msgs::msg::String::SharedPtr msg)
{
	RCLCPP_INFO(this->get_logger(), "Received message: '%s'", msg->data.c_str());
}
void MySubscriber::int_callback(const std_msgs::msg::Int32::SharedPtr msg)
{
	RCLCPP_INFO(this->get_logger(), "Received message: '%d'", msg->data);
}
void MySubscriber::float_callback(const std_msgs::msg::Float32::SharedPtr msg)
{
	RCLCPP_INFO(this->get_logger(), "Received message: '%f'", msg->data);
}
void MySubscriber::bool_callback(const std_msgs::msg::Bool::SharedPtr msg)
{
	RCLCPP_INFO(this->get_logger(), "Received message: '%s'", msg->data ? "true" : "false");
}
	

int main(int argc, char **argv)
{
	rclcpp::init(argc, argv);
	auto node = std::make_shared<MySubscriber>();
	rclcpp::spin(node);
	rclcpp::shutdown();
	return 0;
}